### DISCUSSION TITLE
[Announcement] FAQ for HW2

### DISCUSSION THREAD

**Discussion by sujin0529** [prosyslab-classroom/cs424-program-reasoning#355] [Announcements]
Hi,

We would like to organize some confusing things about the output of Z3 model. (please refer to #47)

The result of your Z3 model and the result of *.expected file do not have to match perfectly.
   - **Okay**: The order of the variables that the Z3 model outputs is different from the expected file.

e.g.
```
satisfiable
 (define-fun tgt_x () Int
   2)
-(define-fun src_x () Int
-  2)
 (define-fun src_y () Int
   1)
+(define-fun src_x () Int
+  2)
```
   - **Okay**: The value of the variable is reasonable, but different from the expected file.

e.g.

**// result of expected file** 
```
satisfiable
(define-fun tgt_x () Int
  2)
(define-fun src_x () Int
  2)
(define-fun src_y () Int
  1)
```

**// assume your result** 
```
satisfiable
(define-fun src_y () Int
  1)
(define-fun tgt_x () Int
  0)
(define-fun src_x () Int
  0)
```
---
If you want to check whether actually passed through test case, you need to modify some things. 

1. please modify your `src/test/dune` according to the following format. (except for sudoku)

**// original**
```
(rule
 (deps pl1.scm)
 (targets pl1.output)
 (action
  (with-stdout-to
   %{targets}
   (run ../pl %{deps}))))

(rule
 (alias runtest)
 (action
  (diff pl1.expected pl1.output)))
```

**// modified**
```
(rule
 (deps pl1.scm)
 (targets pl1.output)
 (action
  (with-stdout-to
   %{targets}
   (pipe-stdout
   (run ../pl %{deps})
   (run head -n 1)))))

(rule
 (alias runtest)
 (action
  (diff pl1.expected pl1.output)))
```

2. Please modify your `dune-project`.

**// original**
`(lang dune 2.3)`

**// modified**
`(lang dune 2.7)`

3. Please leave only the first line of all expected file. (except for sudoku)

**// original**
```
satisfiable
(define-fun x () Int
  2)
(define-fun src_y () Int
  1)
```
**// modified**
```
satisfiable
```

Sorry for the confusion.

**Comment by KihongHeo** [comment]
Again notice that the above changes are NOT mandatory. It is just for your convenience to check the results.
For evaluation, TAs will handle the issues on their own.

**Comment by pingpingy1** [comment]
Not sure if this applies to everyone, but in my case, it seems that there should be exactly one newline character after `satisfiable` or `unsatisfiable` in the `.expected` files.
If you get error messages about newline characters, it might help to check if there are multiple newline characters or no line break after the result.
